### PR TITLE
Implement lakectl local pull

### LINKED ISSUE #6342: Implement lakectl local pull
Parent: https://github.com/treeverse/lakeFS/issues/6239

`lakectl local pull [directory]`

1. Read the references from .lakefs_ref.yaml
2. Perform a remote diff between $at_head and the current HEAD of the branch
3. Perform a local diff (see illustration above) between the specified reference and the current state of the local directory
  a. If there’s any locally uncommitted changes, abort unless --force
5. Apply these changes to local directory
6. Update local .lakefs_ref.yaml setting at_head to the commit ID of HEAD

Flags:

- --presign and --parallelism (see [clone](https://github.com/treeverse/lakeFS/issues/6303) for reference)
- --force or -f - reset any uncommitted local change

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #6342
### Background

Implement lakectl local command which fetches latest changes for a given repository

### New Feature

Part of task https://github.com/treeverse/lakeFS/issues/6239

### Testing Details

Unit testing for tools, lakectl tests will be added for esti upon feature completion. SyncManager will be tested as part of the command (was tested manually)

### Breaking Change?

No

### CODE REVIEW

File: pkg/local/sync.go [hunk 1]
```diff
@@ -228,7 +234,16 @@ func (s *SyncManager) upload(rootPath string, remote *uri.URI, change *Change) e
 }
 
 func (s *SyncManager) deleteLocal(rootPath string, change *Change) error { //nolint:unused
-	panic("Not Implemented")
+	b := s.progressBar.AddSpinner(fmt.Sprintf("delete local path: %s", change.Path))
```
Comment (guy-har) [inline pkg/local/sync.go:237]: Can remove //nolint:unused

File: pkg/local/sync.go [hunk 2]
```diff
@@ -228,7 +234,16 @@ func (s *SyncManager) upload(rootPath string, remote *uri.URI, change *Change) e
...
+	source := filepath.Join(rootPath, change.Path)
+	err := fileutil.RemoveFile(source)
+	if err != nil {
+		return err
+	}
```
Comment (guy-har) [inline pkg/local/sync.go:242]:
```suggestion
	if err != nil {
 	        b.Error()
		return err
	}
```

File: cmd/lakectl/cmd/local_pull.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+		remote, err := idx.GetCurrentURI()
+		if err != nil {
+			DieErr(err)
+		}
+
+		currentBase := remote.WithRef(idx.AtHead)
+		client := getClient()
+		// make sure no local changes
+		localChng := localDiff(cmd.Context(), client, currentBase, idx.LocalPath())
```
Comment (guy-har) [inline cmd/lakectl/cmd/local_pull.go:50]:
```suggestion
		localChange := localDiff(cmd.Context(), client, currentBase, idx.LocalPath())
```

File: cmd/lakectl/cmd/local_pull.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		if len(localChng) > 0 && !force {
+			DieFmt("there are %d uncommitted changes. Either commit them first, or use --force to overwrite local changes",
+				len(localChng))
```
Comment (guy-har) [inline cmd/lakectl/cmd/local_pull.go:53]: The word overwrite might be missleading 
```suggestion
		if len(localChng) > 0 && !force {
			DieFmt("there are %d uncommitted changes. Either commit them first or use --force to revert local changes",
				len(localChng))
```

File: cmd/lakectl/cmd/local_pull.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		}
+
+		// write new index
+		newHead := resolveCommitOrDie(cmd.Context(), client, remote.Repository, remote.Ref)
+		idx, err = local.WriteIndex(idx.LocalPath(), remote, newHead)
+		if err != nil {
+			DieErr(err)
+		}
```
Comment (guy-har) [inline cmd/lakectl/cmd/local_pull.go:61]: I have the same concern here as I had in clone, 
we update the file before succeeding, which my result in corrupted data with no way of knowing. I would expect some way to know that the operation was completed successfully

**Review by guy-har — CHANGES_REQUESTED** [review]
Looks good, minor requests

File: cmd/lakectl/cmd/local_pull.go (hunk 5 above)
Comment (N-o-Z) [reply cmd/lakectl/cmd/local_pull.go:61]: Opened: https://github.com/treeverse/lakeFS/issues/6356

**Review by guy-har — APPROVED** [review]

**Pull request merged**